src/Icicle/Avalanche/Statement/Edsl.hs

```diff
@@ -0,0 +1,150 @@
+-- | Crazy monadic combinators for writing Avalanche programs
+
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Icicle.Avalanche.Statement.Edsl (
+    ValTypeOf, unsafeValType
+  , A, unsafeA
+  , X, unsafeX
+  , S
+  , acc, read, write, slet, sif, whileEq
+  , xnum, xbool
+  , statementOfS
+  ) where
+
+import              Icicle.Common.Base
+import              Icicle.Common.Type
+import              Icicle.Common.Exp
+import              Icicle.Common.Fresh
+
+import              Icicle.Avalanche.Statement.Statement
+import              Icicle.Avalanche.Prim.Flat
+import              Icicle.Avalanche.Annot
+
+import              P
+import              Data.Hashable                  (Hashable)
+
+-- ----------------------------------------
+-- Type synonyms with phantom types
+-- ----------------------------------------
+newtype ValTypeOf t
+ = ValTypeOf ValType
+
+unsafeValType :: ValType -> ValTypeOf t
+unsafeValType = ValTypeOf
+
+
+newtype A n t
+ = A (Name n)
+
+unsafeA :: Name n -> A n t
+unsafeA = A
+
+newtype X n t
+ = X (Exp () n Prim)
+
+unsafeX :: Exp a n Prim -> X n t
+unsafeX = X . reannotX (const ())
+
+
+-- ----------------------------------------
+-- Monads!
+-- The actual magic is a fresh monad with a continuation of what to do with the *rest* of the statement.
+-- ----------------------------------------
```

Comment:
oh, I hadn't thought that far ahead. I was thinking it would just be useful for the definitions of sort and binary search